src/connectStoreController.ts

```diff
@@ -1,126 +1,89 @@
 
   return new Promise((resolve, reject) => {
-    socket.on('connect', () => {
-      const waiters = createWaiters()
-
-      socket.on('message', (message) => {
-        if (message.err) {
-          waiters.reject(message.action, message.err)
-        } else {
-          waiters.resolve(message.action, message.body)
-        }
-      })
-
-      resolve({
-        close: async () => {
-          socket.end()
-        },
-        prune: async () => {
-          socket.sendMessage({
-            action: 'prune',
-          }, (err) => err && console.error(err))
-        },
-        requestPackage: requestPackage.bind(null, socket, waiters),
-        saveState: async () => {
-          socket.sendMessage({
-            action: 'saveState',
-          }, (err) => err && console.error(err))
-        },
-        updateConnections: async (prefix: string, opts: {addDependencies: string[], removeDependencies: string[], prune: boolean}) => {
-          socket.sendMessage({
-            action: 'updateConnections',
-            args: [prefix, opts],
-          }, (err) => err && console.error(err))
-        },
-      })
+    resolve({
+      close: async () => { return },
+      prune: async () => {
+        await limitedRetryFetch(`${remotePrefix}/prune`, {})
+      },
+      requestPackage: requestPackage.bind(null, remotePrefix, limitedRetryFetch),
+      saveState: async () => {
+        await limitedRetryFetch(`${remotePrefix}/saveState`, {})
+      },
+      updateConnections: async (prefix: string, opts: {addDependencies: string[], removeDependencies: string[], prune: boolean}) => {
+        await limitedRetryFetch(`${remotePrefix}/updateConnections`, {
+          opts,
+          prefix,
+        })
+      },
     })
   })
 }
 
-function createWaiters () {
-  const waiters = {}
-  return {
-    add (id: string) {
-      waiters[id] = deffered()
-      return waiters[id].promise
-    },
-    resolve (id: string, obj: object) {
-      if (waiters[id]) {
-        waiters[id].resolve(obj)
-        delete waiters[id]
-      }
-    },
-    reject (id: string, err: object) {
-      if (waiters[id]) {
-        waiters[id].reject(err)
-        delete waiters[id]
-      }
-    },
-  }
-}
-
-// tslint:disable-next-line
-function noop () {}
-
-function deffered<T> (): {
-  promise: Promise<T>,
-  resolve: (v: T) => void,
-  reject: (err: Error) => void,
-} {
-  let pResolve: (v: T) => void = noop
-  let pReject: (err: Error) => void = noop
-  const promise = new Promise<T>((resolve, reject) => {
-    pResolve = resolve
-    pReject = reject
+function retryFetch (limit: (fn: () => PromiseLike<object>) => Promise<object>, url: string, body: object): Promise<object> { // tslint:disable-line
+  return limit(async () => {
+    const response = await got(url, {
+      body: JSON.stringify(body),
+      headers: { 'Content-Type': 'application/json' },
+      method: 'POST',
+    })
+    return JSON.parse(response.body)
+  }).catch((e) => {
```

Comment:
you can use async/await syntax here with a regular try/catch

also, I think the try/catch should be inside the function and cover just the `got()` call, as errors by `JSON.parse` will always be passed through